File: packages/clients/ios/Sources/DoubloonChecker/SolanaChecker.swift
import Foundation
import CryptoKit

/// Lightweight Solana entitlement checker using direct JSON-RPC.
///
/// Uses Foundation's URLSession for networking and CryptoKit for SHA-256.
/// No dependency on any Solana SDK.
///
/// ```swift
/// let checker = SolanaChecker(
///     rpcUrl: URL(string: "https://api.mainnet-beta.solana.com")!,
///     programId: "Dub1oon11111111111111111111111111111111111"
/// )
///
/// let result = try await checker.checkEntitlement(
///     productId: "a7f3c9...",
///     wallet: "UserWalletBase58..."
/// )
///
/// if result.entitled {
///     // Grant access
/// }
/// ```
public actor SolanaChecker {
    private let rpcUrl: URL
    private let programId: String
    private let commitment: String
    private let rpc: RpcClient

    public init(rpcUrl: URL, programId: String, commitment: String = "confirmed") {
        self.rpcUrl = rpcUrl
        self.programId = programId
        self.commitment = commitment
        self.rpc = RpcClient()
    }

    /// Check if a wallet holds an active entitlement for the given product.
    public func checkEntitlement(productId: String, wallet: String) async throws -> EntitlementCheck {
        let entitlement = try await getEntitlement(productId: productId, wallet: wallet)
        return DoubloonChecker.checkEntitlement(entitlement)
    }

    /// Fetch the raw entitlement data, or nil if not found.
    public func getEntitlement(productId: String, wallet: String) async throws -> Entitlement? {
        let address = try SolanaPDA.deriveEntitlementAddress(
            productIdHex: productId,
            userWalletBase58: wallet,
            programId: programId
        )

        guard let data = try await fetchAccountData(address: address) else {
            return nil
        }

        return try SolanaDeserializer.deserializeEntitlement(data: data)
    }

    /// Fetch product metadata, or nil if not found.
    public func getProduct(productId: String) async throws -> Product? {
        let address = try SolanaPDA.deriveProductAddress(
            productIdHex: productId,
            programId: programId
        )

        guard let data = try await fetchAccountData(address: address) else {
            return nil
        }

        return try SolanaDeserializer.deserializeProduct(data: data)
    }

    /// Batch check multiple products for one wallet.
    public func checkEntitlements(
        productIds: [String],
        wallet: String
    ) async throws -> [String: EntitlementCheck] {
        let addresses = try productIds.map { pid in
            try SolanaPDA.deriveEntitlementAddress(
                productIdHex: pid,
                userWalletBase58: wallet,
                programId: programId
            )
        }

        let result: SolanaMultipleAccountsResult = try await rpc.call(
            url: rpcUrl,
            method: "getMultipleAccounts",
            params: [addresses, ["encoding": "base64", "commitment": commitment]]
        )

        var checks: [String: EntitlementCheck] = [:]
        for (i, pid) in productIds.enumerated() {
            let entitlement: Entitlement?
            if let account = result.value[i],
               let base64Data = account.data.first,
               let data = Data(base64Encoded: base64Data) {
                entitlement = try SolanaDeserializer.deserializeEntitlement(data: data)
            } else {
                entitlement = nil
            }
            checks[pid] = DoubloonChecker.checkEntitlement(entitlement)
        }

        return checks
    }

    // MARK: - Private

    private func fetchAccountData(address: String) async throws -> Data? {
        let result: SolanaAccountResult = try await rpc.call(
            url: rpcUrl,
            method: "getAccountInfo",
            params: [address, ["encoding": "base64", "commitment": commitment]]
        )

        guard let account = result.value,
              let base64Data = account.data.first,
              let data = Data(base64Encoded: base64Data) else {
            return nil
        }

        return data
    }
}

// MARK: - RPC Response Types

private struct SolanaAccountResult: Decodable {
    let value: SolanaAccountValue?
}

private struct SolanaAccountValue: Decodable {
    let data: [String]
    let lamports: UInt64
    let owner: String
}

private struct SolanaMultipleAccountsResult: Decodable {
    let value: [SolanaAccountValue?]
}

// MARK: - PDA Derivation

enum SolanaPDA {
    private static let base58Alphabet = Array("123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz")

    static func base58Decode(_ string: String) throws -> Data {
        var bytes: [UInt8] = [0]
        for char in string {
            guard let idx = base58Alphabet.firstIndex(of: char) else {
                throw NSError(domain: "DoubloonChecker", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid base58 character: \(char)"])
            }
            var carry = base58Alphabet.distance(from: base58Alphabet.startIndex, to: idx)
            for j in 0..<bytes.count {
                carry += Int(bytes[j]) * 58
                bytes[j] = UInt8(carry & 0xFF)
                carry >>= 8
            }
            while carry > 0 {
                bytes.append(UInt8(carry & 0xFF))
                carry >>= 8
            }
        }
        // Leading zeros
        for char in string {
            if char != "1" { break }
            bytes.append(0)
        }
        return Data(bytes.reversed())
    }

    static func base58Encode(_ data: Data) -> String {
        var digits: [Int] = [0]
        for byte in data {
            var carry = Int(byte)
            for j in 0..<digits.count {
                carry += digits[j] << 8
                digits[j] = carry % 58
                carry /= 58
            }
            while carry > 0 {
                digits.append(carry % 58)
                carry /= 58
            }
        }
        var result = ""
        for byte in data {
            if byte != 0 { break }
            result.append("1")
        }
        for i in stride(from: digits.count - 1, through: 0, by: -1) {
            result.append(base58Alphabet[digits[i]])
        }
        return result
    }

    static func hexToData(_ hex: String) -> Data {
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let nextIndex = hex.index(index, offsetBy: 2)
            let byteString = hex[index..<nextIndex]
            data.append(UInt8(byteString, radix: 16)!)
            index = nextIndex
        }
        return data
    }

    /// Find a Solana Program Derived Address.
    static func findProgramAddress(seeds: [Data], programId: String) throws -> (String, UInt8) {
        let programIdData = try base58Decode(programId)
        let marker = "ProgramDerivedAddress".data(using: .utf8)!

        for bump: UInt8 in stride(from: 255, through: 0, by: -1) {
            var hashInput = Data()
            for seed in seeds {
                hashInput.append(seed)
            }
            hashInput.append(bump)
            hashInput.append(programIdData)
            hashInput.append(marker)

            let hash = SHA256.hash(data: hashInput)
            let hashData = Data(hash)

            // A valid PDA must NOT be on the ed25519 curve.
            // Simplified check: try to interpret as a compressed point and reject if valid.
            // For a robust check we'd need full ed25519 point decompression.
            // The probability of a SHA-256 output being on-curve is ~2^-128, so in practice
            // the first bump (255) almost always works.
            if !isOnEd25519Curve(hashData) {
                return (base58Encode(hashData), bump)
            }
        }

        throw NSError(domain: "DoubloonChecker", code: 2, userInfo: [NSLocalizedDescriptionKey: "Could not find PDA"])
    }

    /// Derive the entitlement PDA address.
    static func deriveEntitlementAddress(
        productIdHex: String,
        userWalletBase58: String,
        programId: String
    ) throws -> String {
        let seeds: [Data] = [
            "entitlement".data(using: .utf8)!,
            hexToData(productIdHex),
            try base58Decode(userWalletBase58),
        ]
        let (address, _) = try findProgramAddress(seeds: seeds, programId: programId)
        return address
    }

    /// Derive the product PDA address.
    static func deriveProductAddress(productIdHex: String, programId: String) throws -> String {
        let seeds: [Data] = [
            "product".data(using: .utf8)!,
            hexToData(productIdHex),
        ]
        let (address, _) = try findProgramAddress(seeds: seeds, programId: programId)
        return address
    }

    /// Derive a product ID hex from a slug (SHA-256).
    static func deriveProductIdHex(slug: String) -> String {
        let hash = SHA256.hash(data: slug.data(using: .utf8)!)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }

    // Simplified ed25519 on-curve check.
    // Full check requires point decompression; this uses a heuristic.
    // In practice, SHA-256 outputs are almost never valid curve points.
    private static func isOnEd25519Curve(_ data: Data) -> Bool {
        // The ed25519 curve has order ~2^252.6. A random 256-bit value has
        // probability ~0.5 * 2^-3 of being a valid y-coordinate, and even then
        // the x-coordinate recovery can fail. For PDA derivation, the Solana
        // runtime does a full decompression check.
        //
        // Here we do a conservative fast-path: if the top 3 bits indicate
        // the value exceeds the field prime p = 2^255 - 19, it's definitely
        // not on the curve. Otherwise, assume it might be (try next bump).
        // This is a safe overapproximation — we may try a few extra bumps
        // but will never accept an invalid PDA.
        guard data.count == 32 else { return true }
        let topByte = data[31] & 0x7F // clear sign bit
        if topByte > 0x7E { return false } // > p, definitely not on curve
        // Conservative: assume it could be on curve
        return true
    }
}

// MARK: - Deserialization

enum SolanaDeserializer {
    static func deserializeEntitlement(data: Data) throws -> Entitlement {
        var offset = 8 // skip Anchor discriminator
        let productId = readProductId(data, &offset)
        let user = readPubkey(data, &offset)
        let grantedAt = readI64(data, &offset)
        let expiresAt = readI64(data, &offset)
        let autoRenew = readBool(data, &offset)
        let source = readU8(data, &offset)
        let sourceId = readString(data, &offset)
        let active = readBool(data, &offset)
        let revokedAt = readI64(data, &offset)
        let revokedBy = readPubkey(data, &offset)

        let defaultPubkey = "11111111111111111111111111111111"

        return Entitlement(
            productId: productId,
            user: user,
            grantedAt: Date(timeIntervalSince1970: TimeInterval(grantedAt)),
            expiresAt: expiresAt == 0 ? nil : Date(timeIntervalSince1970: TimeInterval(expiresAt)),
            autoRenew: autoRenew,
            source: EntitlementSource(rawValue: source) ?? .platform,
            sourceId: sourceId,
            active: active,
            revokedAt: revokedAt == 0 ? nil : Date(timeIntervalSince1970: TimeInterval(revokedAt)),
            revokedBy: revokedBy == defaultPubkey ? nil : revokedBy
        )
    }

    static func deserializeProduct(data: Data) throws -> Product {
        var offset = 8
        let creator = readPubkey(data, &offset)
        let productId = readProductId(data, &offset)
        let name = readString(data, &offset)
        let metadataUri = readString(data, &offset)
        let createdAt = readI64(data, &offset)
        let updatedAt = readI64(data, &offset)
        let active = readBool(data, &offset)
        let frozen = readBool(data, &offset)
        let entitlementCount = readU64(data, &offset)
        let delegateCount = readU16(data, &offset)
        let defaultDuration = readI64(data, &offset)

        return Product(
            creator: creator,
            productId: productId,
            name: name,
            metadataUri: metadataUri,
            createdAt: Date(timeIntervalSince1970: TimeInterval(createdAt)),
            updatedAt: Date(timeIntervalSince1970: TimeInterval(updatedAt)),
            active: active,
            frozen: frozen,
            entitlementCount: UInt64(entitlementCount),
            delegateCount: UInt16(delegateCount),
            defaultDuration: Int64(defaultDuration)
        )
    }

    // MARK: - Readers

    private static func readU8(_ data: Data, _ offset: inout Int) -> UInt8 {
        let val = data[offset]
        offset += 1
        return val
    }

    private static func readBool(_ data: Data, _ offset: inout Int) -> Bool {
        let val = data[offset] != 0
        offset += 1
        return val
    }

    private static func readU16(_ data: Data, _ offset: inout Int) -> Int {
        let val = Int(data[offset]) | (Int(data[offset + 1]) << 8)
        offset += 2
        return val
    }

    private static func readU32(_ data: Data, _ offset: inout Int) -> Int {
        let val = Int(data[offset])
            | (Int(data[offset + 1]) << 8)
            | (Int(data[offset + 2]) << 16)
            | (Int(data[offset + 3]) << 24)
        offset += 4
        return val
    }

    private static func readI64(_ data: Data, _ offset: inout Int) -> Int64 {
        var val: Int64 = 0
        for i in 0..<8 {
            val |= Int64(data[offset + i]) << (i * 8)
        }
        offset += 8
        return val
    }

    private static func readU64(_ data: Data, _ offset: inout Int) -> Int64 {
        var val: UInt64 = 0
        for i in 0..<8 {
            val |= UInt64(data[offset + i]) << (i * 8)
        }
        offset += 8
        return Int64(val)
    }

    private static func readPubkey(_ data: Data, _ offset: inout Int) -> String {
        let bytes = data[offset..<(offset + 32)]
        offset += 32
        return SolanaPDA.base58Encode(Data(bytes))
    }

    private static func readProductId(_ data: Data, _ offset: inout Int) -> String {
        let bytes = data[offset..<(offset + 32)]
        offset += 32
        return bytes.map { String(format: "%02x", $0) }.joined()
    }

    private static func readString(_ data: Data, _ offset: inout Int) -> String {
        let len = readU32(data, &offset)
        let strData = data[offset..<(offset + len)]
        offset += len
        return String(data: strData, encoding: .utf8) ?? ""
    }
}

File: packages/clients/android/src/main/kotlin/com/doubloon/checker/SolanaChecker.kt
package com.doubloon.checker

import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest
import java.util.Date

/**
 * Lightweight Solana entitlement checker using direct JSON-RPC.
 *
 * Uses OkHttp for networking and java.security for SHA-256.
 * No dependency on any Solana SDK.
 *
 * ```kotlin
 * val checker = SolanaChecker(
 *     rpcUrl = "https://api.mainnet-beta.solana.com",
 *     programId = "Dub1oon11111111111111111111111111111111111"
 * )
 *
 * val result = checker.checkEntitlement("a7f3c9...", "UserWalletBase58...")
 * if (result.entitled) {
 *     // Grant access
 * }
 * ```
 */
class SolanaChecker(
    private val rpcUrl: String,
    private val programId: String,
    private val commitment: String = "confirmed",
    private val rpc: RpcClient = RpcClient(),
) {
    /**
     * Check if a wallet holds an active entitlement.
     */
    suspend fun checkEntitlement(productId: String, wallet: String): EntitlementCheck {
        val entitlement = getEntitlement(productId, wallet)
        return com.doubloon.checker.checkEntitlement(entitlement)
    }

    /**
     * Fetch the raw entitlement data, or null if not found.
     */
    suspend fun getEntitlement(productId: String, wallet: String): Entitlement? {
        val address = SolanaPda.deriveEntitlementAddress(productId, wallet, programId)
        val data = fetchAccountData(address) ?: return null
        return SolanaDeserializer.deserializeEntitlement(data)
    }

    /**
     * Fetch product metadata, or null if not found.
     */
    suspend fun getProduct(productId: String): Product? {
        val address = SolanaPda.deriveProductAddress(productId, programId)
        val data = fetchAccountData(address) ?: return null
        return SolanaDeserializer.deserializeProduct(data)
    }

    /**
     * Batch check multiple products for one wallet.
     */
    suspend fun checkEntitlements(
        productIds: List<String>,
        wallet: String,
    ): Map<String, EntitlementCheck> {
        val addresses = productIds.map { pid ->
            SolanaPda.deriveEntitlementAddress(pid, wallet, programId)
        }

        val params = JSONArray().apply {
            put(JSONArray(addresses))
            put(JSONObject().apply {
                put("encoding", "base64")
                put("commitment", commitment)
            })
        }

        val json = rpc.call(rpcUrl, "getMultipleAccounts", params)
        val values = json.getJSONObject("result").getJSONArray("value")

        return productIds.mapIndexed { i, pid ->
            val entitlement = if (!values.isNull(i)) {
                val account = values.getJSONObject(i)
                val dataArray = account.getJSONArray("data")
                val base64Data = dataArray.getString(0)
                val bytes = android.util.Base64.decode(base64Data, android.util.Base64.DEFAULT)
                SolanaDeserializer.deserializeEntitlement(bytes)
            } else {
                null
            }
            pid to com.doubloon.checker.checkEntitlement(entitlement)
        }.toMap()
    }

    private suspend fun fetchAccountData(address: String): ByteArray? {
        val params = JSONArray().apply {
            put(address)
            put(JSONObject().apply {
                put("encoding", "base64")
                put("commitment", commitment)
            })
        }

        val json = rpc.call(rpcUrl, "getAccountInfo", params)
        val result = json.getJSONObject("result")

        if (result.isNull("value")) return null
        val value = result.getJSONObject("value")
        val dataArray = value.getJSONArray("data")
        val base64Data = dataArray.getString(0)

        return android.util.Base64.decode(base64Data, android.util.Base64.DEFAULT)
    }
}

// MARK: - PDA Derivation

object SolanaPda {
    private const val BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

    fun base58Decode(input: String): ByteArray {
        val bytes = mutableListOf<Int>(0)
        for (char in input) {
            val idx = BASE58_ALPHABET.indexOf(char)
            require(idx >= 0) { "Invalid base58 character: $char" }
            var carry = idx
            for (j in bytes.indices) {
                carry += bytes[j] * 58
                bytes[j] = carry and 0xFF
                carry = carry shr 8
            }
            while (carry > 0) {
                bytes.add(carry and 0xFF)
                carry = carry shr 8
            }
        }
        // Leading zeros
        for (char in input) {
            if (char != '1') break
            bytes.add(0)
        }
        return bytes.reversed().map { it.toByte() }.toByteArray()
    }

    fun base58Encode(bytes: ByteArray): String {
        val digits = mutableListOf(0)
        for (byte in bytes) {
            var carry = byte.toInt() and 0xFF
            for (j in digits.indices) {
                carry += digits[j] shl 8
                digits[j] = carry % 58
                carry /= 58
            }
            while (carry > 0) {
                digits.add(carry % 58)
                carry /= 58
            }
        }
        val sb = StringBuilder()
        for (byte in bytes) {
            if (byte.toInt() != 0) break
            sb.append('1')
        }
        for (i in digits.indices.reversed()) {
            sb.append(BASE58_ALPHABET[digits[i]])
        }
        return sb.toString()
    }

    fun hexToBytes(hex: String): ByteArray {
        return ByteArray(hex.length / 2) { i ->
            hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
    }

    fun bytesToHex(bytes: ByteArray): String {
        return bytes.joinToString("") { "%02x".format(it) }
    }

    /**
     * Find a Solana Program Derived Address.
     */
    fun findProgramAddress(seeds: List<ByteArray>, programId: String): Pair<String, Int> {
        val programIdBytes = base58Decode(programId)
        val marker = "ProgramDerivedAddress".toByteArray(Charsets.UTF_8)

        for (bump in 255 downTo 0) {
            val digest = MessageDigest.getInstance("SHA-256")
            for (seed in seeds) {
                digest.update(seed)
            }
            digest.update(bump.toByte())
            digest.update(programIdBytes)
            digest.update(marker)
            val hash = digest.digest()

            // PDA must not be on the ed25519 curve.
            // SHA-256 outputs are overwhelmingly unlikely to be valid curve points.
            // Conservative check: reject values >= field prime p = 2^255 - 19.
            if (!isLikelyOnCurve(hash)) {
                return Pair(base58Encode(hash), bump)
            }
        }

        throw IllegalStateException("Could not find PDA")
    }

    fun deriveEntitlementAddress(productIdHex: String, walletBase58: String, programId: String): String {
        val seeds = listOf(
            "entitlement".toByteArray(Charsets.UTF_8),
            hexToBytes(productIdHex),
            base58Decode(walletBase58),
        )
        return findProgramAddress(seeds, programId).first
    }

    fun deriveProductAddress(productIdHex: String, programId: String): String {
        val seeds = listOf(
            "product".toByteArray(Charsets.UTF_8),
            hexToBytes(productIdHex),
        )
        return findProgramAddress(seeds, programId).first
    }

    fun deriveProductIdHex(slug: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        return bytesToHex(digest.digest(slug.toByteArray(Charsets.UTF_8)))
    }

    private fun isLikelyOnCurve(data: ByteArray): Boolean {
        if (data.size != 32) return true
        val topByte = data[31].toInt() and 0x7F
        if (topByte > 0x7E) return false
        // Conservative: assume could be on curve
        return true
    }
}

// MARK: - Deserialization

object SolanaDeserializer {
    fun deserializeEntitlement(data: ByteArray): Entitlement {
        var offset = 8 // skip Anchor discriminator
        val productId = readProductId(data, offset); offset += 32
        val user = readPubkey(data, offset); offset += 32
        val grantedAt = readI64(data, offset); offset += 8
        val expiresAt = readI64(data, offset); offset += 8
        val autoRenew = readBool(data, offset); offset += 1
        val source = readU8(data, offset); offset += 1
        val (sourceId, sourceIdLen) = readString(data, offset); offset += sourceIdLen
        val active = readBool(data, offset); offset += 1
        val revokedAt = readI64(data, offset); offset += 8
        val revokedBy = readPubkey(data, offset); offset += 32

        val defaultPubkey = "11111111111111111111111111111111"

        return Entitlement(
            productId = productId,
            user = user,
            grantedAt = Date(grantedAt * 1000),
            expiresAt = if (expiresAt == 0L) null else Date(expiresAt * 1000),
            autoRenew = autoRenew,
            source = EntitlementSource.fromValue(source),
            sourceId = sourceId,
            active = active,
            revokedAt = if (revokedAt == 0L) null else Date(revokedAt * 1000),
            revokedBy = if (revokedBy == defaultPubkey) null else revokedBy,
        )
    }

    fun deserializeProduct(data: ByteArray): Product {
        var offset = 8
        val creator = readPubkey(data, offset); offset += 32
        val productId = readProductId(data, offset); offset += 32
        val (name, nameLen) = readString(data, offset); offset += nameLen
        val (metadataUri, metadataUriLen) = readString(data, offset); offset += metadataUriLen
        val createdAt = readI64(data, offset); offset += 8
        val updatedAt = readI64(data, offset); offset += 8
        val active = readBool(data, offset); offset += 1
        val frozen = readBool(data, offset); offset += 1
        val entitlementCount = readU64(data, offset); offset += 8
        val delegateCount = readU16(data, offset); offset += 2
        val defaultDuration = readI64(data, offset); offset += 8

        return Product(
            creator = creator,
            productId = productId,
            name = name,
            metadataUri = metadataUri,
            createdAt = Date(createdAt * 1000),
            updatedAt = Date(updatedAt * 1000),
            active = active,
            frozen = frozen,
            entitlementCount = entitlementCount,
            delegateCount = delegateCount,
            defaultDuration = defaultDuration,
        )
    }

    private fun readU8(data: ByteArray, offset: Int): Int = data[offset].toInt() and 0xFF

    private fun readBool(data: ByteArray, offset: Int): Boolean = data[offset].toInt() != 0

    private fun readU16(data: ByteArray, offset: Int): Int =
        (data[offset].toInt() and 0xFF) or ((data[offset + 1].toInt() and 0xFF) shl 8)

    private fun readU32(data: ByteArray, offset: Int): Int =
        (data[offset].toInt() and 0xFF) or
            ((data[offset + 1].toInt() and 0xFF) shl 8) or
            ((data[offset + 2].toInt() and 0xFF) shl 16) or
            ((data[offset + 3].toInt() and 0xFF) shl 24)

    private fun readI64(data: ByteArray, offset: Int): Long {
        var value = 0L
        for (i in 0 until 8) {
            value = value or ((data[offset + i].toLong() and 0xFF) shl (i * 8))
        }
        return value
    }

    private fun readU64(data: ByteArray, offset: Int): Long = readI64(data, offset)

    private fun readPubkey(data: ByteArray, offset: Int): String =
        SolanaPda.base58Encode(data.sliceArray(offset until offset + 32))

    private fun readProductId(data: ByteArray, offset: Int): String =
        SolanaPda.bytesToHex(data.sliceArray(offset until offset + 32))

    private fun readString(data: ByteArray, offset: Int): Pair<String, Int> {
        val len = readU32(data, offset)
        val str = String(data, offset + 4, len, Charsets.UTF_8)
        return Pair(str, 4 + len)
    }
}
